File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int m=grid.length;
        int n=grid[0].length;
        int fresh=0;
        int minutes=0;
         Queue<int[]> q=new LinkedList<>();
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                    if(grid[i][j]==2){q.offer(new int[]{i,j});}
                    if(grid[i][j]==1){fresh++;}
            }
        }
       
        while(!q.isEmpty() && fresh>0){
            int size=q.size();
            minutes++;
            for(int k=0;k<size;k++){
                int[] curr=q.poll();
                int i=curr[0];
                int j=curr[1];
                if((i-1)>=0 && grid[i-1][j]==1){ grid[i-1][j]=2 ; q.offer(new int[]{i-1,j}); fresh--;}
                if((i+1)<m && grid[i+1][j]==1){ grid[i+1][j]=2 ; q.offer(new int[]{i+1,j}); fresh--; }
                if((j-1)>=0 && grid[i][j-1]==1){ grid[i][j-1]=2 ; q.offer(new int[]{i,j-1}); fresh--;}
                if((j+1)<n && grid[i][j+1]==1){grid[i][j+1]=2 ; q.offer(new int[]{i,j+1}); fresh--;}
            }
        }
        if(fresh!=0){return -1;}
        return minutes;
    }
}

File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution(object):
    directions=[(1,0),(-1,0),(0,1),(0,-1)]
    def orangesRotting(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m=len(grid)
        n=len(grid[0])
        fresh=0
        minutes=0
        q=deque()

        for i in range(m):
            for j in range(n):
                if grid[i][j]==1:
                    fresh+=1
                if grid[i][j]==2:
                    q.append((i,j))
        
        while (q and fresh>0):
            size=len(q)
            for i in range(size):
              r,c=q.popleft()
              for dx,dy in self.directions:
                nr=r+dx
                nc=c+dy
                if (nr>=0 and nr<m and nc>=0 and nc<n):
                    if(grid[nr][nc]==1):
                        q.append((nr,nc))
                        grid[nr][nc]=2
                        fresh-=1
            minutes+=1
        if (fresh==0):
            return minutes
        return -1


        